### PR TITLE
Replace getElementType with getValueType for MAP in AvroGenericRecord…

### LINKED ISSUE #29516: [Bug]: org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.toProtoValue fails to handle MAP type schema
### What happened?

Beam Version - 2.52.0
Language - Java

This is the problematic function:
Screenshot 2023-11-22 at 12 09 26

In line 278 is the case handling a MAP Avro schema type. In line 280 it is trying to compute the value type of the Map but accessing the `Schema::getElementType` method, which is implemented only by the `org.Apache.Avro.ArraySchema` (and not by the `org.Apache.Avro.MapSchema`). This is causing the case to throw an `AvroRuntimeException` with a message "Not an array: ...", and NEVER succeed the handling.

I believe that in line 280, instead of calling `Schema::getElementType`, it should call `Schema::getValueType` which is implemented by the `org.Apache.Avro.MapSchema` class.

P.S - It seems that same bug exists in the org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.fieldDescriptorFromAvroField function when handling a MAP type schema.

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### LINKED ISSUE #31652: [Bug]: `getElementType` method in `AvroGenericRecordToStorageApiProto.java` causes AvroRuntimeException
### What happened?

Hi,

While trying the BigQuery Storage Write Api, I've come across an issue in the `AvroGenericRecordToStorageApiProto.java` file in the Apache Beam repository. Specifically, the issue lies in the following lines of code:

- [Line 250](https://github.com/apache/beam/blame/561bf2f07e15719e9d698d2d57455e050246105e/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java#L250)
- [Line 349](https://github.com/apache/beam/blame/561bf2f07e15719e9d698d2d57455e050246105e/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java#L349)

In these lines, the `getElementType` method is being called on a `schema` object:

```java
Schema valueType = TypeWithNullability.create(schema.getElementType()).getType();
```

However, the `schema` object is of type `MapSchema`, which does not implement the `getElementType` method. This method is only implemented for `ArraySchema`. For all other types, it defaults to the `Schema` class and throws an `AvroRuntimeException`:

```java
/** If this is an array, returns its element type. */
public Schema getElementType() {
  throw new AvroRuntimeException("Not an array: " + this);
}
```

Instead of calling `getElementType`,  `getValueType` could be called as follows:

```java
Schema valueType = TypeWithNullability.create(schema.getValueType()).getType();
```

I believe this change could potentially resolve the issue. I look forward to your feedback.

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by codertimu**
Fixes #31652 
Fixes #29516 

While trying the BigQuery Storage Write Api, I've come across an issue in the `AvroGenericRecordToStorageApiProto.java` file in the Apache Beam repository. Specifically, the issue lies in the following lines of code:

- [Line 250](https://github.com/apache/beam/blame/561bf2f07e15719e9d698d2d57455e050246105e/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java#L250)
- [Line 349](https://github.com/apache/beam/blame/561bf2f07e15719e9d698d2d57455e050246105e/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java#L349)

In these lines, the `getElementType` method is being called on a `schema` object:

```java
Schema valueType = TypeWithNullability.create(schema.getElementType()).getType();
```

However, the `schema` object is of type `MapSchema`, which does not implement the `getElementType` method. This method is only implemented for `ArraySchema`. For all other types, it defaults to the `Schema` class and throws an `AvroRuntimeException`:

```java
/** If this is an array, returns its element type. */
public Schema getElementType() {
  throw new AvroRuntimeException("Not an array: " + this);
}
```

Instead of calling `getElementType`,  `getValueType` could be called as follows:

```java
Schema valueType = TypeWithNullability.create(schema.getValueType()).getType();
```

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by liferoad** [comment]
Please fix the splotless issues. You could do `./gradlew spotlessApply`.

**Comment by codertimu** [comment]
assign set of reviewers

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java [hunk 1]
```diff
@@ -266,6 +268,28 @@ enum TestEnum {
...
+            .type(BASE_SCHEMA)
+            .name("aMap")
+            .type()
+            .map()
+            .values()
+            .stringType()
+            .mapDefault(defaultVal)
+            .endRecord();
+  }
+
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java:292]:
```suggestion
  private static final Schema SCHEMA_WITH_MAP =
      SchemaBuilder.record("TestMap")
          .fields()
          .name("nested")
          .type()
          .optional()
          .type(BASE_SCHEMA)
          .name("aMap")
          .type()
          .map()
          .values()
          .stringType()
          .mapDefault(ImmutableMap.<String, Object>builder().put("key1", "value1").build())
          .endRecord();
```

Can we do this all inline instead?

**Review by ahmedabu98 — COMMENTED** [review]
Looks great! Thanks for fixing this.

Left one nit

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java (hunk 1 above)
Comment (codertimu) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java:292]: sure

**Review by ahmedabu98 — APPROVED** [review]

**Pull request merged**